biocrnpyler/mechanism.py

```diff
@@ -331,11 +332,111 @@ def update_reactions(self, s1, s2, component = None, kb = None, ku = None, part_
             ku = component.get_parameter("ku", part_id = part_id, mechanism = self)
         if component == None and (kb == None or ku == None):
             raise ValueError("Must pass in a Component or values for kb, ku.")
-
-        complex = ComplexSpecies([s1, s2])
+        if(complex==None):
+            complex = ComplexSpecies([s1, s2])
         rxns = [Reaction([s1, s2], [complex], k=kb, k_rev=ku)]
         return rxns
 
+class Combinatorial_Cooperative_Binding(Mechanism):
+    """a reaction where some number of binders bind combinatorially to a bindee"""
+    def __init__(self,name="Combinatorial_Cooperative_binding",
+                            mechanism_type="cooperative_binding"):
+        Mechanism.__init__(self,name,mechanism_type)
+    def make_cooperative_complex(self,combo,bindee,cooperativity):
+        """given a list of binders and their cooperativities, make a complex
+        that contains the binders present N number of times where N is
+        each one's cooperativity"""
+        complexed_species_list = []
+        for binder in combo:
+            binder_cooperativity = int(cooperativity[binder.name])
+            #I hope that cooperativity is an int! what if it isn't
+            if(binder_cooperativity > 1):
+                complexed_species_list += [Multimer(binder,binder_cooperativity)]
+            else:
+                complexed_species_list += [binder]
+        complexed_species_list += [bindee]
+        if(len(complexed_species_list)==1):
+            myspecies = complexed_species_list[0]
+        else:
+            myspecies = ComplexSpecies(complexed_species_list)
+        return myspecies
+    def update_species(self,binders,bindee,cooperativity=None,\
+                            component = None, part_id = None, **kwords):
+        cooperativity_dict = {}
+        for binder in binders:
+            binder_partid = part_id+"_"+binder.name
+            if ((cooperativity == None) or (type(cooperativity)==dict and binder_partid not in cooperativity) \
+                                                                                        and (component != None)):
```

Comment:
some here as well: component is not None

Reply:
ok I will fix it in my code but it seems like this is present in other mechanisms too

Reply:
I see I'll try to weed them out from the other places.